Adding Enforcement Attributes to Demand Data

jacobmalleau opened the pull request:
Adding the ability to handle enforcement related events in the CDS Curb Events and Metrics data.  Based on work with LADOT in the SMART Grant Collab.
Resolves #173 
Resolves #174

jacobmalleau committed:
Adding Enforcement Attributes to Demand Data

Adding the ability to handle enforcement related events in the CDS Curb Events and Metrics data. 
Based on work with LADOT in the SMART Grant Collab

jacobmalleau committed:
Enforcement Metric - Total Events

Addition of metric to handle summation of events

schnuerle commented:
We will be talking about this at today's public [Curb Working Group meeting](https://github.com/openmobilityfoundation/curb-data-specification/wiki/Web-Conference-2025.04.22-Curb).

jacobmalleau committed:
Updates

Slight edits to enforcement type names

schnuerle commented:
From our [Curb Working Group](https://github.com/openmobilityfoundation/curb-data-specification/wiki/Web-Conference-2025.06.17-Curb) meeting today:

- Could add an 'enforcement' array key value object that can be part of Events to be captured: muni code, violation name, enforcement action taken, fee/cost, ticket id
- Also applies to curb objects, how to connect to these objects?
- Minneapolis and San Joes support this

SFMTA pushes our parking citation data to a public open data feed, and it would be great to have this information available in CDS events. 
Current feed: https://data.sfgov.org/Transportation/SFMTA-Parking-Citations-Fines/ab4h-6ztd/about_data

LADOT does as well: https://data.lacity.org/Transportation/Parking-Citations/4f5p-udkv/about_data

jacobmalleau committed:
Add connecting events to Curb Objects

Also slight edits to language.

jacobmalleau commented:
Added a commit to connect Events to Objects, but held off on adding the key-value pairs as I want to see if anyone else has thoughts on attributes to add:
- muni code
- violation name
- enforcement action taken
- fee/cost
- ticket id
- Others?

schnuerle committed:
Merge branch 'dev' into jacobmalleau-enforcementimprovements

schnuerle commented:
> ... held off on adding the key-value pairs as I want to see if anyone else has thoughts on attributes to add:

Where would these key value pairs be added?

schnuerle commented:
@jacobmalleau I think go ahead and add a new Enforcement data type with these fields, like we [created for External References](https://github.com/openmobilityfoundation/curb-data-specification/blob/dev/data-types.md). Then we can talk about what enforcement fields to add/remove/edit beyond your short list.

You can then add the new enforcement data object as a field in Events, like with [External References in Events](https://github.com/openmobilityfoundation/curb-data-specification/tree/dev/events#curb-event) (bottom of table) and Curbs.

jacobmalleau committed:
Adding external reference to enforcement info.

jacobmalleau committed:
Adding Enforcement Reference in Data Types

Will likely need a review from City staff and/or enforcement vendors to validate fields and whether additional fields are needed.

schnuerle committed:
Merge branch 'dev' into jacobmalleau-enforcementimprovements

schnuerle committed:
Fixed typo

schnuerle committed:
Order and formatting

schnuerle commented:
Thanks, looks good, I made some formatting updates, and changed it to just 'Enforcement'. The 'External Reference' name was really just because it was a reference to an external thing.

schnuerle committed:
Add violation_end event type

schnuerle committed:
Added an `enforcement_id`

schnuerle committed:
Noted `violation_end` enforcement type

schnuerle committed:
Merge branch 'dev' into jacobmalleau-enforcementimprovements

schnuerle commented:
We will be reviewing this at [tomorrow's public Curb Working Group meeting](https://github.com/openmobilityfoundation/curb-data-specification/wiki/Web-Conference-2025.08.12-Curb). However we won't have much time to discuss, so leave your comments and suggestions here as general or inline comments to spark changes before the [CDS 1.1 release](https://github.com/openmobilityfoundation/curb-data-specification/wiki/Release-1.1.0) is made.

JohnLundstrom-Minneapolis commented:
> Added a commit to connect Events to Objects, but held off on adding the key-value pairs as I want to see if anyone else has thoughts on attributes to add:
> 
> * muni code
> * violation name
> * enforcement action taken
> * fee/cost
> * ticket id
> * Others?

I'd recommend the following fields, dependent upon how common they are in other citation systems:

*Status (for us this has been a two character field for the primary entry, such as "IS" for issued, "VA" for void approved, and  "WR" for warning).

*Plate AND State if transmitting such information

Finally, how are we handling the location in this concept?

bhamlinSDOT commented:
> > Added a commit to connect Events to Objects, but held off on adding the key-value pairs as I want to see if anyone else has thoughts on attributes to add:
> > 
> > * muni code
> > * violation name
> > * enforcement action taken
> > * fee/cost
> > * ticket id
> > * Others?
> 
> I'd recommend the following fields, dependent upon how common they are in other citation systems:
> 
> *Status (for us this has been a two character field for the primary entry, such as "IS" for issued, "VA" for void approved, and "WR" for warning).
> 
> *Plate AND State if transmitting such information
> 
> Finally, how are we handling the location in this concept?

Agree w/ @JohnLundstrom-Minneapolis that Status would be helpful. 

For location - our citation data is currently lat/long coords.

We also have date/time that the citation was written. I assume the date/time + location (if not associated with a curb_zone_id) would be captured in event_location and event_time?

schnuerle commented:
@JohnLundstrom-Minneapolis : @bhamlinSDOT is right. Date, time, and location and other fields are captured in the general Event data. 

>  I assume the date/time + location (if not associated with a curb_zone_id) would be captured in event_location and event_time?

MichaelWDanko commented on events/README.md:
```diff
@@ -199,6 +200,10 @@ Curb Event Type `event_type` enumerates the set of possible types of Curb Event.
 | `scheduled_report` | event source reported status at a scheduled interval |
 | `enter_area`       | vehicle enters the relevant geographic area |
 | `exit_area`        | vehicle exits the relevant geographic area |
+| `vehicle_detected` | the event of detecting or locating a vehicle at the curb. For example, from a manual surveying or license plate recognition machine |
+| `vehicle_violation` | the event of detecting a vehicle that is doing something illegal or not allowed at the curb. For example, a non permitted vehicle is parked or someone overstayed the time limit. This does not reflect the actual ticket given. Typically requires a `vehicle_detected` event first  |
+| `citation_given`   | the event of providing a ticket or citation to a vehicle. Typically requires a `vehicle_violation` event first |
+| `violation_end`    | the end of a violation, when it is no longer an issue | 
```
The event types `vehicle_violation`, `violation_end`, and `citation_given` have been renamed to `vehicle_violation_start`, `vehicle_violation_end`, and `citation_issued` respectively. These changes:

1. Clarify that the violation events relate specifically to vehicle compliance violations.
2. Improve consistency and precision in naming conventions.
3. Replace “given” with “issued” to reflect the formal creation of a citation, regardless of whether it has been physically delivered or received.
4. Specify that `vehicle_violation_end` applies to time-based violations where an end can be programmatically detected.

```suggestion
| `vehicle_detected` | detection or observation of a vehicle within a curb area. Can originate from manual surveying, automated license plate recognition (LPR) systems, or other detection methods. |
| `vehicle_violation_start` | start of a compliance violation at a curb location, triggered when a vehicle is not permitted or has exceeded allowed time limits. This event may be published after detecting a vehicle but does not require a `vehicle_detected` event. |
| `vehicle_violation_end` | resolution of a compliance violation, when the vehicle is no longer in violation of regulations. Used for time-based violations where the end of the violation can be detected. Requires a preceding vehicle_violation_start event for the same vehicle. | 
| `citation_issued` | issuance of a ticket or citation to a vehicle. May be published in addition to a related vehicle_violation_start event. |
```

michael-danko-passport replied:
The event types `vehicle_violation`, `violation_end`, and `citation_given` have been renamed to `vehicle_violation_start`, `vehicle_violation_end`, and `citation_issued` respectively. These changes:

1. Clarify that the violation events relate specifically to vehicle compliance violations.
2. Improve consistency and precision in naming conventions.
3. Replace “given” with “issued” to reflect the formal creation of a citation, regardless of whether it has been physically delivered or received.
4. Specify that `vehicle_violation_end` applies to time-based violations where an end can be programmatically detected.

```suggestion
| `vehicle_detected` | detection or observation of a vehicle within a curb area. Can originate from manual surveying, automated license plate recognition (LPR) systems, or other detection methods. |
| `vehicle_violation_start` | start of a compliance violation at a curb location, triggered when a vehicle is not permitted or has exceeded allowed time limits. This event may be published after detecting a vehicle but does not require a `vehicle_detected` event. |
| `vehicle_violation_end` | resolution of a compliance violation, when the vehicle is no longer in violation of regulations. Used for time-based violations where the end of the violation can be detected. Requires a preceding vehicle_violation_start event for the same vehicle. | 
| `citation_issued` | issuance of a ticket or citation to a vehicle. May be published in addition to a related vehicle_violation_start event. |
```

michael-danko-passport commented on events/README.md:
```diff
@@ -180,6 +180,7 @@ A Curb Event is represented as a JSON object, whose fields are as follows:
 | `vehicle_blocked_lane_types` | Array of [Lane Type](#lane-type) | Conditionally Required | Type(s) of lane blocked by the vehicle performing the event. If no lanes are blocked by the vehicle performing the event, the array should be empty.  Required for sources capable of determining it for the following event_types: _park_start_ |
 | `curb_occupants` | Array of [Curb Occupant](#curb-occupants) | Conditionally Required | Current occupants of the Curb Zone. If the sensor is capable of identifying the linear location of the vehicle, then elements are sorted in ascending order according to the start property of the linear reference. Otherwise, elements appear in no particular order. Required for sources capable of determining i ...
 | `actual_cost` | Integer | Optional | If available from the source, the actual cost, in the currency defined in currency, paid by the curb user for this event. The currency type is sent in with the [REST Endpoints](#rest-endpoints) JSON object. All costs should be given as integers in the currency's smallest unit. As an example, to represent $1 USD, specify an amount of 100 (for 100 cents). |
+| `enforcement` | Array of [Enforcement][enforcement] objects | Optional | Enforcement information related to this Curb Event, relevant to the moment in time the event happens. Only used for enforcement related events such as `vehicle_detected`, `vehicle_violation`, `citation_given`, and `violation_end`. |
```
These changes are to align with proposed changes down below where we define the new event types.

```suggestion
| `enforcement` | Array of [Enforcement][enforcement] objects | Optional | Enforcement information related to this Curb Event, relevant to the moment in time the event happens. Only used for enforcement related events such as `vehicle_detected`, `vehicle_violation_start`, `vehicle_violation_end`, and `citation_issued`. |
```

michael-danko-passport commented on events/README.md:
```diff
@@ -160,8 +160,8 @@ A Curb Event is represented as a JSON object, whose fields are as follows:
 | `curb_area_ids` | Array of [UUID][uuid] | Conditionally Required | Unique IDs of the Curb Area where the event occurred. Since Curb Areas can overlap, an event may happen in more than one. Required for events that occurred in a known Curb Area, if known and used, for these event_types: _enter_area, exit_area, park_start, park_end_ |
-| `curb_space_id` | [UUID][uuid] | Conditionally Required | Unique ID of the Curb Space where the event occurred. Required for events that occurred at a known Curb Space, if known and used, for these event_types: _park_start, park_end, enter_area, exit_area_ |
-| `curb_object_id` | [UUID][uuid] | Conditionally Required | Unique ID of the Curb Object where the event occurred. Required for events that occurred at a known Curb Object, if known and used, for these event_types: _park_start, park_end, enter_area, exit_area_ |
+| `curb_space_id` | [UUID][uuid] | Conditionally Required | Unique ID of the Curb Space where the event occurred. Required for events that occurred at a known Curb Space, if known and used, for these event_types: _park_start, park_end, enter_area, exit_area, vehicle_detected, vehicle_violation, citation_given_ |
+| `curb_object_id` | [UUID][uuid] | Conditionally Required | Unique ID of the Curb Object where the event occurred. Required for events that occurred at a known Curb Object, if known and used, for these event_types: _park_start, park_end, enter_area, exit_area, vehicle_detected, vehicle_violation, citation_given_ |
```
These changes are to align with proposed changes down below where we define the new event types.

```suggestion
| `curb_space_id` | [UUID][uuid] | Conditionally Required | Unique ID of the Curb Space where the event occurred. Required for events that occurred at a known Curb Space, if known and used, for ALL _event_types_. |
| `curb_object_id` | [UUID][uuid] | Conditionally Required | Unique ID of the Curb Object where the event occurred. Required for events that occurred at a known Curb Object, if known and used, for ALL _event_types_. |
```

michael-danko-passport replied:
I don't believe these are breaking changes since we're saying that if we know and use the space/object ID for any event, it would add more data that can be used/consumed. Current implementation conditionally limits for certain event types and I'm not sure there's much value.

michael-danko-passport commented on events/README.md:
```diff
@@ -160,8 +160,8 @@ A Curb Event is represented as a JSON object, whose fields are as follows:
 | `event_session_id` | [UUID][uuid] | Optional | May be provided to tie known connected `park_start` and `park_end` event types together by a unique session ID. If _not_ confident of being able to determine a `park_end` event at some time after `park_start` is recorded (i.e., you cannot detect when a vehicle departs), then do _not_ use session_id. This field may be most useful to payment companie ...
 | `curb_zone_id` | [UUID][uuid] | Conditionally Required | Unique ID of the Curb Zone where the event occurred. Required for events that occurred at a known Curb Zone for ALL _event_types_. |
 | `curb_area_ids` | Array of [UUID][uuid] | Conditionally Required | Unique IDs of the Curb Area where the event occurred. Since Curb Areas can overlap, an event may happen in more than one. Required for events that occurred in a known Curb Area, if known and used, for these event_types: _enter_area, exit_area, park_start, park_end_ |
```
```suggestion
| `curb_zone_id` | [UUID][uuid] | Conditionally Required | Unique ID of the Curb Zone where the event occurred. Required for events that occurred in a known Curb Zone, if known and used, for ALL _event_types_. |
| `curb_area_ids` | Array of [UUID][uuid] | Conditionally Required | Unique IDs of the Curb Area where the event occurred. Since Curb Areas can overlap, an event may happen in more than one. Required for events that occurred in a known Curb Area, if known and used, for ALL _event_types_. |
```

michael-danko-passport replied:
Similar to below, I don't believe these are breaking changes since we're saying that if we know and use the space/object ID for any event, it would add more data that can be used/consumed.

michael-danko-passport commented on data-types.md:
```diff
@@ -4,21 +4,39 @@ This CDS data types page catalogs the data objects (fields, types, requirements,
+| `ticket_id`      | String | Optional            | The unique id that represents the ticket being given. |
+| `name`           | String | Optional            | Name of the rule being enforced or citation being given. |
+| `action_taken`   | String | Optional            | What action was taken to enforce the rule being violated. Typical well-known values are `ticket_served`, `ticket_posted`, `ticket_registered`, or `ticket_emailed`. |
+| `ticket_cost`    | String | Optional            | The cost associated with the given violation/ticket issued. |
```
Since violations often escalate/change over a lifecycle, do we want to specify the original cost? I think this is more or less a change in description to indicate the cost at issuance.

```suggestion
| `ticket_cost`    | String | Optional            | The original cost associated with the given violation/ticket issued. |
```

michael-danko-passport commented on data-types.md:
```diff
@@ -4,21 +4,39 @@ This CDS data types page catalogs the data objects (fields, types, requirements,
+| `municipal_code` | string | Optional            | The unique code created by the municipality or enforcement agency to identify the type of rule being enforced. |
+| `ticket_id`      | String | Optional            | The unique id that represents the ticket being given. |
+| `name`           | String | Optional            | Name of the rule being enforced or citation being given. |
+| `action_taken`   | String | Optional            | What action was taken to enforce the rule being violated. Typical well-known values are `ticket_served`, `ticket_posted`, `ticket_registered`, or `ticket_emailed`. |
```
Do we need to define an enumeration of values that can be used/expected in the `action_taken`?

schnuerle replied:
I feel like we should keep this free form text for now. I think there are too many options to try and enumerate at this point, even if we only list well known values and allow the creation of others.  Maybe we can work to create a well known list in a new issue.

michael-danko-passport commented on data-types.md:
```diff
@@ -4,21 +4,39 @@ This CDS data types page catalogs the data objects (fields, types, requirements,
+| `enforcement_id` | UUID   | Optional            | An optional unique identifer of an enforcement incident, generated the first time an enforcement event is recorded, and referenced in future enforcement events. | 
+| `municipal_code` | string | Optional            | The unique code created by the municipality or enforcement agency to identify the type of rule being enforced. |
+| `ticket_id`      | String | Optional            | The unique id that represents the ticket being given. |
+| `name`           | String | Optional            | Name of the rule being enforced or citation being given. |
```
In the lifecycle of a citation/ticket, `name` could be a common used name. Should this clarify that it's the name of the violation that required an enforcement action?
```suggestion
| `violation_name`           | String | Optional            | Name of the rule being enforced or citation being given. |
```

michael-danko-passport commented:
There are a few places where we are using `citation`, `ticket`, and `violation` and I think we should be consistent throughout the spec. I would propose the following:
- `citation` -> Replaces the usage of `ticket` and represents a single instance of a broken infraction.
- `violation` -> The city or municipal code that was violated  

If we can get consensus in one direction or another there are some additional changes that will need to be made.

schnuerle committed:
Note original cost

Co-authored-by: michael-danko-passport <99669313+michael-danko-passport@users.noreply.github.com>

schnuerle committed:
Clarified field with `violation_name` renaming

Co-authored-by: michael-danko-passport <99669313+michael-danko-passport@users.noreply.github.com>

schnuerle committed:
Noted that zones and areas can use any event type

Co-authored-by: michael-danko-passport <99669313+michael-danko-passport@users.noreply.github.com>

schnuerle committed:
Allows use of any event type if known and used

Reduces need to enumerate event types.

Co-authored-by: michael-danko-passport <99669313+michael-danko-passport@users.noreply.github.com>

schnuerle committed:
Renaming enforcement fields for consistency

Co-authored-by: michael-danko-passport <99669313+michael-danko-passport@users.noreply.github.com>

schnuerle committed:
Enforcement fields edits for consistency, clarity

Co-authored-by: michael-danko-passport <99669313+michael-danko-passport@users.noreply.github.com>

schnuerle committed:
Minor formatting

schnuerle committed:
Minor description updates for event types

schnuerle commented:
> There are a few places where we are using `citation`, `ticket`, and `violation` and I think we should be consistent throughout the spec. I would propose the following:
> 
> * `citation` -> Replaces the usage of `ticket` and represents a single instance of a broken infraction.
> * `violation` -> The city or municipal code that was violated
> 
> If we can get consensus in one direction or another there are some additional changes that will need to be made.

Are you thinking here, for ticket_id and ticket_cost? 

https://github.com/openmobilityfoundation/curb-data-specification/blob/jacobmalleau-enforcementimprovements/data-types.md#enforcement

Is a ticket different than a citation? Can there be a ticket ID and a citation ID too, or are those really always the same thing?

schnuerle commented:
@JohnLundstrom-Minneapolis and @bhamlinSDOT can you propose a definition of a `status` field? 

It should be an open ended 'string' type, so you can use two letter abbreviations if needed, or other cities can use what they need.

kenyaw commented:
> > There are a few places where we are using `citation`, `ticket`, and `violation` and I think we should be consistent throughout the spec. I would propose the following:
> > 
> > * `citation` -> Replaces the usage of `ticket` and represents a single instance of a broken infraction.
> > * `violation` -> The city or municipal code that was violated
> > 
> > If we can get consensus in one direction or another there are some additional changes that will need to be made.
> 
> Are you thinking here, for ticket_id and ticket_cost?
> 
> https://github.com/openmobilityfoundation/curb-data-specification/blob/jacobmalleau-enforcementimprovements/data-types.md#enforcement
> 
> Is a ticket different than a citation? Can there be a ticket ID and a citation ID too, or are those really always the same thing?

@schnuerle  In San Francisco, we utilize `citation` for parking violations, so the change from `ticket` seems fine. 

Regarding the definition for `violation`,  cities enforce infractions from various levels of government. A more descriptive definition statement could be "The city/municipal, _county, state, or provincial code_ that was violated."

schnuerle committed:
Merge branch 'dev' into jacobmalleau-enforcementimprovements

schnuerle commented:
Based on @kenyaw's comments, I think you can align things to citation, and away from ticket @michael-danko-passport. And update the violation description to "The city/municipal, county, state, provincial, or federal code that was violated." Note I added federal to that list.

schnuerle committed:
Updated terms from ticket to citation, for consistency

JohnLundstrom-Minneapolis commented:
> @JohnLundstrom-Minneapolis and @bhamlinSDOT can you propose a definition of a `status` field?
> 
> It should be an open ended 'string' type, so you can use two letter abbreviations if needed, or other cities can use what they need.

I'm realizing that we wouldn't be capturing changes to the status over time (such as when a citation is cancelled on some future date), as those wouldn't be distinct curb events, but we still have some options at the time the citation is issued, mainly delineating between a warning and an active citation. My first attempt is:

"Describes the disposition of the citation at the time it was issued which can be used to identify alternate use cases such as warnings."

schnuerle committed:
Update field name in Metrics to citation_issued

michael-danko-passport committed:
Update README.md

Documentation updates:
- Clarifies distinction between vehicle detection and enter/exit event types.
- Text formatting changes

michael-danko-passport committed:
Incorporates changes from Slack discussion.

Applies changes discussed via Slack with members from WGSC.
- Documentation updates to the descriptions of vehicle_detected (new) and enter_area (existing) events to distinguish between these.
- Adds a new `enforcement.is_warning` field
- Adds a new `enforcement.violations` array
- Defines a new `Violations` object
- Moves violation specific fields to the new violations array

michael-danko-passport committed:
Change citation_id from UUID to String

michael-danko-passport committed:
Changes `enforcement` key in CDS Event to be an object

- Previously was an array of Enforcement objects
- Change aligns with a single enforcement object for each event, with the Enforcement object now containing an array of violations

schnuerle committed:
Violations TOC and hierarchy

michael-danko-passport committed:
Adds Enforcement and Violations data objects to TOC

schnuerle committed:
Removed data type links from TOC

schnuerle committed:
Minor cleanup

schnuerle committed:
Merge pull request #199 from openmobilityfoundation/feature-enforcement-object-changes

Incorporate Enforcement updates from WGSC discussions

schnuerle reviewed (approved):
Thanks to the OMF member WGSC and public WG participants for the work and clarity on this feature!
